### PR TITLE
feat: Add S3 utility

### PR DESCRIPTION
**Pull request by arihantsurana**
### What changes are proposed in this PR?
---
* Added basic s3 utility methods
* Modified AwsConnection to contain both get_resource and get_client methods
* paused code coverage merging / unittests

### How was this tested?
---
* Added Unittests

### CODE REVIEW

File: hip_data_tools/aws/s3.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+from joblib import load
+
+
+class S3Util:
+    def __init__(self, conn, bucket):
+        self.conn = conn
+        self.bucket = bucket
+
+    def _client(self):
```
Comment (vbmudalige) [inline hip_data_tools/aws/s3.py]: In Java, we define public methods before private methods as a practice. The normal order is public, protected, default, private. We can use the same for python as well.

File: tests/aws/test_s3.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        cls.sample_file_content = str(uuid.uuid4())
+        with open(cls.sample_file_location, 'w+') as f:
+            f.write(cls.sample_file_content)
+
+    @classmethod
+    def tearDownClass(cls):
+        os.remove(cls.sample_file_location)
+
+    @mock_s3
+    def test_file_upload(self):
```
Comment (vbmudalige) [inline tests/aws/test_s3.py]: We can consider using some consistent unit test naming standard. [https://dzone.com/articles/7-popular-unit-test-naming](https://dzone.com/articles/7-popular-unit-test-naming)
I prefer to use "Should_ExpectedBehavior_When_StateUnderTest" convention. It is up to you to decide is this something we should consider going forward.

File: tests/aws/test_s3.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        bucket = "TEST_BUCKET"
+        conn = AwsConnection(mode="standard_env_var", region_name="ap-southeast-2", settings={})
+        s3u = S3Util(conn=conn, bucket=bucket)
+        s3u.create_bucket()
+        upload_key = "temp.txt"
+        s3u.upload_file(local_file_path=self.sample_file_location, s3_key=upload_key)
+        redown_file = self.sample_file_location + "re"
```
Comment (vbmudalige) [inline tests/aws/test_s3.py]: I prefer to use format strings over concatenation. It's more readable.

Comment (vbmudalige) [inline tests/aws/test_s3.py]:
```suggestion
        redown_file = "{}re".format(self.sample_file_location)
```

File: tests/aws/test_s3.py (hunk 2 above)
Comment (arihantsurana) [reply tests/aws/test_s3.py]: yes we should. I'll use this one.

File: hip_data_tools/aws/s3.py (hunk 1 above)
Comment (arihantsurana) [reply hip_data_tools/aws/s3.py]: done

File: tests/aws/test_s3.py (hunk 3 above)
Comment (arihantsurana) [reply tests/aws/test_s3.py]: good point.

**Comment by Gurado** [comment]
no ticket link in PR @arihantsurana

**Comment by arihantsurana** [comment]
> no ticket link in PR @arihantsurana

@Gurado this repo is opensource, we don't reference our internal tickets here, we have GitHub issues for bugs.

File: hip_data_tools/aws/s3.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    def download_object_and_deserialse(self, s3_key, local_file_path=None):
+        """
+        Download a serialised object from S3 and deserialse
+        Args:
+            s3_key (string): Absolute path on s3 to the file
+            local_file_path (string): The deserialsed object
+        Returns: object
+        """
+        if local_file_path is None:
+            local_file_path = "/tmp/tmp_file" + str(uuid.uuid4())
```
Comment (vbmudalige) [inline hip_data_tools/aws/s3.py]:
```suggestion
            local_file_path = "/tmp/tmp_file{}".format(str(uuid.uuid4()))
```

File: tests/aws/test_s3.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        cls.sample_file_content = str(uuid.uuid4())
+        with open(cls.sample_file_location, 'w+') as f:
+            f.write(cls.sample_file_content)
+
+    @classmethod
+    def tearDownClass(cls):
+        os.remove(cls.sample_file_location)
+
+    @mock_s3
+    def test_should__upload_then_download_file_from_s3__when_using_s3util(self):
```
Comment (vbmudalige) [inline tests/aws/test_s3.py]: I think the word "test_" is redundant

Comment (vbmudalige) [inline tests/aws/test_s3.py]:
```suggestion
    def Should__uploadThenDownloadFileFromS3__When_UsingS3util(self):
```

Comment (vbmudalige) [reply tests/aws/test_s3.py]: We don't have to tightly strict with python naming standards as there is no reusability requirement for the test methods. The only goal of the test method name is just readability.

Comment (arihantsurana) [reply tests/aws/test_s3.py]: I don't want to pollute the namespace with camelcase. it also causes the linter to complain. I'd rather keep it with python standard.

Comment (vbmudalige) [reply tests/aws/test_s3.py]: Make sense. Let's proceed with this standard then.

Comment (arihantsurana) [reply tests/aws/test_s3.py]: this is required by the unittest framework to discover tests

**Pull request merged**